bookshelf/api_v3/ec2.py

```diff
@@ -0,0 +1,321 @@
+            instance.modify_attribute('BlockDeviceMapping',
+                                      ["%s=%d" % (dev, 1)])
+
+    # add a tag to our instance
+    connection.create_tags([instance.id], tags)
+
+    if log:
+        log_green("Public dns: %s" % instance.public_dns_name)
+
+    # returns our new instance
+    return instance
+
+
+@implementer(ICloudInstance)
+@provider(ICloudInstanceFactory)
+class EC2():
+    """
+    Class representing an EC2 instance, that provides methods for interacting
+    with the instance.
+
+    :ivar connection: A boto connection object to interact with ec2.
+    :ivar instance: A boto instance object.
+    :ivar config: An EC2Configuration describing the configuration for this EC2
+        instance.
+    :ivar state: An EC2State describing this EC2 instance.
+    """
+    def __init__(self, instance, connection, config, state):
+        self.connection = connection
+        self.instance = instance
+        self.config = config
+        self.state = state
+
+    cloud_type = u'ec2'
+
+    @property
+    def ip_address(self):
+        return self.instance.ip_address
+
+    @property
+    def image_basename(self):
+        return self.config.image_basename
+
+    @property
+    def username(self):
+        return self.config.username
+
+    @property
+    def key_filename(self):
+        return self.config.key_filename
+
+    @property
+    def distro(self):
+        return self.state.distro
+
+    @property
+    def region(self):
+        return self.state.region
+
+    @classmethod
+    def create_from_config(cls, config, distro, region):
+        parsed_config = EC2Configuration.create(config)
+        connection = _connect_to_ec2(
+            region=region,
+            credentials=parsed_config.credentials
+        )
+        instance = _create_server_ec2(
+            connection=connection,
+            region=region,
+            disk_name=parsed_config.disk_name,
+            disk_size=parsed_config.disk_size,
+            ami=parsed_config.ami,
+            key_pair=parsed_config.key_pair,
+            instance_type=parsed_config.instance_type,
+            tags=parsed_config.tags,
+            security_groups=parsed_config.security_groups,
+            delete_on_termination=True,
+            log=False,
+            wait_for_ssh_available=True
+        )
+        state = EC2State(
+            instance_id=instance.id,
+            region=region,
+            distro=distro,
+        )
+
+        return cls(
+            connection=connection,
+            instance=instance,
+            config=parsed_config,
+            state=state
+        )
+
+    @classmethod
+    def create_from_saved_state(cls, config, saved_state):
+        parsed_config = EC2Configuration.create(config)
+        state = EC2State.create(saved_state)
+        connection = _connect_to_ec2(
+            region=state.region,
+            credentials=parsed_config.credentials
+        )
+
+        timeout = 600
```

Comment:
Sure, although we didn't bother adding it to the interface, so it only works for EC2 for now.